test_conformance/subgroups/workgroup_common_kernels.cpp

```diff
@@ -0,0 +1,104 @@
+//
+// Copyright (c) 2020 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "workgroup_common_kernels.h"
+
+const char* bcast_source = "__kernel void test_bcast(const __global Type *in, "
+                           "__global int4 *xy, __global Type *out)\n"
+                           "{\n"
+                           "    int gid = get_global_id(0);\n"
+                           "    XY(xy,gid);\n"
+                           "    Type x = in[gid];\n"
+                           "    out[gid] = sub_group_broadcast(x, xy[gid].z);\n"
+
+                           "}\n";
+
+const char* redadd_source = "__kernel void test_redadd(const __global Type "
+                            "*in, __global int4 *xy, __global Type *out)\n"
+                            "{\n"
+                            "    int gid = get_global_id(0);\n"
+                            "    XY(xy,gid);\n"
+                            "    out[gid] = sub_group_reduce_add(in[gid]);\n"
```

Comment:
Do we / should we check the type of the result of the reduction, anywhere?  Also, do we / should we check the type of the operand, too?

Reply:
Each kernel defines which data type will be used using [define](https://github.com/gwawiork/OpenCL-CTS/blob/extsub_2711_2020_all_changes/test_conformance/subgroups/subhelpers.h#L1285)
Validation is done on host side (at least results). 
I do not know if on device side there is a such need. Have we ever done such checks on device side ?